File: backend/src/main/java/com/codementor/backend/service/ProblemService.java
package com.codementor.backend.service;

import com.codementor.backend.dto.ProblemProgressResponse;
import com.codementor.backend.dto.ProblemRequest;
import com.codementor.backend.dto.ProblemResponse;
import com.codementor.backend.dto.ProblemStatusFilter;
import com.codementor.backend.entity.Difficulty;
import org.springframework.data.domain.Page;

import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ProblemService {

    ProblemResponse createProblem(ProblemRequest request);

    List<ProblemResponse> getAllProblems();

    ProblemResponse getProblemById(Long id);

    Page<ProblemResponse> searchProblems(
            String title,
            int page,
            int size
    );

    Page<ProblemResponse> filterProblems(
            String title,
            Difficulty difficulty,
            int page,
            int size
    );

    void deleteProblem(Long id);

    ProblemResponse updateProblem(
            Long id,
            ProblemRequest request
    );

    List<ProblemResponse> getProblemsByDifficulty(
            Difficulty difficulty
    );

        ProblemProgressResponse getMyProblemProgress(
                String userEmail
        );

        List<Long> getMySolvedProblemIds(
                String userEmail
        );

        Page<ProblemResponse> filterProblemsForUser(

                String userEmail,

                String title,

                Difficulty difficulty,

                ProblemStatusFilter status,

                Pageable pageable
        );
}

File: frontend/src/services/problemService.ts
import api from "./api";

import type {
  Difficulty,
  Problem,
  ProblemPage,
  ProblemProgress,
  ProblemRequest,
} from "../types/problem";


export type ProblemStatus =
  | "ALL"
  | "SOLVED"
  | "UNSOLVED";


export const getAllProblems = async (): Promise<Problem[]> => {

  const response =
    await api.get<Problem[]>("/problems");

  return response.data;
};

export const createProblem = async (
  request: ProblemRequest
): Promise<Problem> => {
  const response = await api.post<Problem>(
    "/problems",
    request
  );

  return response.data;
};


export const updateProblem = async (
  id: number,
  request: ProblemRequest
): Promise<Problem> => {
  const response = await api.put<Problem>(
    `/problems/${id}`,
    request
  );

  return response.data;
};


export const deleteProblem = async (
  id: number
): Promise<void> => {
  await api.delete(`/problems/${id}`);
};

export const getProblemById = async (
  id: number
): Promise<Problem> => {

  const response =
    await api.get<Problem>(
      `/problems/${id}`
    );

  return response.data;
};


export const filterProblems = async (

  title: string,

  difficulty: Difficulty | "",

  page: number,

  size: number

): Promise<ProblemPage> => {

  const response =
    await api.get<ProblemPage>(
      "/problems/filter",
      {
        params: {

          title,

          difficulty:
            difficulty || undefined,

          page,

          size,
        },
      }
    );

  return response.data;
};


export const filterMyProblems = async (

  title: string,

  difficulty: Difficulty | "",

  status: ProblemStatus,

  page: number,

  size: number

): Promise<ProblemPage> => {

  const response =
    await api.get<ProblemPage>(
      "/problems/filter/me",
      {
        params: {

          title,

          difficulty:
            difficulty || undefined,

          status,

          page,

          size,
        },
      }
    );

  return response.data;
};

export const getMyProblemProgress =
  async (): Promise<ProblemProgress> => {

    const response =
      await api.get<ProblemProgress>(
        "/problems/progress/me"
      );

    return response.data;
  };


export const getMySolvedProblemIds = async (): Promise<number[]> => {
  const response = await api.get<number[]>(
    "/problems/solved/me"
  );

  return response.data;
};

